### ISSUE TITLE
Implement Directional Animation

### ISSUE THREAD

**Issue by controllerface** [controllerface/bvge#39]
Now that basic skeletal animation is working, it is time to move forward and work on animations that are a little more like what a real game would have (though still probably very rough).

Because the game world is 2D, most/all animations will need to be created with the character facing one way, most likely to the right. And then when the game is running, the animation should be flipped along the Y axis if the player is facing to the left. 

I am not 100% sure how this will work, but because I am not doing back-face culling, and the animations are _technically_ in 3D, it should be possible to just flip the animations along that vertical axis. One nice thing that might arise from this is that left and right graphics for arms and legs should actually stay on the correct, though they will of course be mirrored. This shouldn't be an issue though, and I don't plan to have any text or other things that would look odd if mirrored this way.

It is probably best to keep with the existing humanoid model for these tests, just to avoid adding extra working with supporting more models. I will start off with just an "idle" animation but I hope to also get a simple walk and/or run animation working as well if I can swing it.

**Comment by controllerface** [comment]
Doing a little bit of research into ways I might implement this (before doing any real work yet) my initial thought was to use a matrix, similarly to how the animation kernels work, perhaps even combined with them. This would probably look similar to that logic, and implemented as an additional step in those kernels. 

After considering this for a bit though, I think it may be easier to implement a separate kernel that performs a basic reflection along the desired axis, relative to the centroid of the root hull of the armature. Since this information is readily available, and it would remove the need for another expensive `float16` object. 

I am hoping that either of these methods won't interfere with the collision calculations, I have no reason think they will, but this is new territory for me, and the effect here will be to mirror the hull points, hopefully this doesn't end up requiring extra work to keep the physics calculations stable.

If I get stuck, there is always "Plan B" which is to basically double the animations in blender, making every other animation index a mirror of the previous by rotating the entire model, but that is not what i want to do if I can avoid it.

**Comment by controllerface** [comment]
I ended up working on this a bit sort of on accident, and have now added a new humanoid model. The new blender project has a better layout, where the character's limbs are animated in 3D, so it will work more naturally with most common animation workflows. Because of this, I may actually be able to simply mirror all animations and get left/right facing without the need for runtime math. Because I can embed multiple UV sets, I can also switch texture when moving from left to right facing, making the issue with backwards text go away. 

I still have some work to do to make sure the new layout renders properly, since I ma using 3D to my advantage, I will need to do depth testing, but that should be very easy. 

Now... for that walk cycle animation..

**Comment by controllerface** [comment]
First attempt at a walk cycle is done, and I did some cleanup on the vertex weights so it doesn't even look half bad!

I now have a pretty decent workflow for character building, which is super handy. For smaller characters, like bugs and small animals, I think the turn around time will be pretty quick indeed. But that is getting ahead of myself a little bit...

Now that I can make decent animations, I need to actually structure how animations will work, as right now just whatever the first animation is within the model file is played on a loop. I need to define the set of all animations I will need for a main character, and also what will be required for NPCs, etc. I also need to test how flipping right to left is going to work. 

In order to make the character limbs render correctly, I used the depth buffer and I set the value of the `z` vector component manually to 1, 0 or -1 depending on if the hull is closer, in between, or further away respectively. I technically have a lot more "layers" I could play with, but for now this is fine enough.

Overall really happy with progress, there's still quite a bit of work to do, but having decent animations feels like a pretty decent level-up for the project.

**Comment by controllerface** [comment]
I think it is time to start working on a state machine for the player. Looking at all that is required to be handled, it's at the stage where a formal structure is needed.

I have been weighing which side to implement, Java or Open CL, and I am not 100% sure which to choose. I think longer term, I may be better off with the logic inside kernels, but it may be easier to conceptualize things in a Java class. The main issue just becomes the data passing from the GPU to the CPU. That's a big argument in favor of Open CL, simply because it will reduce the amount data moving back and forth each frame.

**Comment by controllerface** [comment]
After much tinkering and "side-quests", I now have directional animation working!

In the end, i took the approach of simply rotating the model transform by 180 degrees for left-facing animations. This ends up being the best of both worlds for the two other approaches I though I might take. 

Originally, I figured I would need to flip things in the animations directly, so all the bones would have to be effectively rotated, which kept me away from that option at first. Then I did look into flipping them in blender and just having 2 directions for every animation. It would work, but it would get clunky and require double updates every time some animation was tweaked, and double the number of state transitions. 

When i looked again, I realized that just the model transform, which occurs are the "root" of the animation tree traversal was all that actually needed to be changed. And the change is a simple matrix multiply that flips the direction the model is facing before all the bone animations are applied. The nice thing is, this is exactly the same as having the duplicate animations, it's not flipping all the bones, just the root transform _before_ the bones. In hindsight, I should have assumed this was possible, I did know I'd need to deal with matrices, but kind of overthought it, assuming they would need to be applied at every step in the chain.

I still need to do some work to flip the z-buffer data when the model is facing the opposite direction, so that the correct limb is closer to the camera, as well as set up a second UV map for this direction. And to test things work as I expect, I should probably write some small, but legible text on the character's arms and/or legs so I can tell for sure there's no issues with mirroring. 

Having this in place certainly is one more step toward this feeling like an actual game-like thing. Looking froward to refining this further and getting into some actual game mechanics soon!
